### PR TITLE
iOS crosscompilation under linux

### PR DESCRIPTION
**Pull request by radikalliberal**
**Description**:
added support for crosscompilation under linux
the used toolchain was generated with: 
[https://github.com/tpoechtrager/cctools-port](https://github.com/tpoechtrager/cctools-port)

the toolchain has the following configuration
clang-8.0.1
arm64
iOS-12.2

[link ](https://drive.google.com/file/d/1-vURXcolFPu_lSXsRdND2R3r8Ehmqaz2/view?usp=sharing) to prebuild toolchain

**onnx also had to be changed!**

```
 diff --git a/onnx/proto_utils.h b/onnx/proto_utils.h
 
 index ec125fde..3de807ac 100644
 --- a/onnx/proto_utils.h
 +++ b/onnx/proto_utils.h
 @@ -31,7 +31,7 @@ bool ParseProtoFromBytes(Proto* proto, const char* buffer, size_t length) {
    // respectively.
    ::google::protobuf::io::ArrayInputStream input_stream(buffer, static_cast<int>(length));
    ::google::protobuf::io::CodedInputStream coded_stream(&input_stream);
 -  coded_stream.SetTotalBytesLimit((2048LL << 20) - 1, 512LL << 20);
 +  coded_stream.SetTotalBytesLimit((2048LL << 20) - 1);
    return proto->ParseFromCodedStream(&coded_stream);
  }
```

SetTotalBytesLimit only allows one argument for protobuf 3.11.2

call to build.py:
export PATH=~/clang-8.0.1_arm64_iOS-12.2/glibc-2.17_x86_64/bin:$PATH

python3 tools/ci_build/build.py \
        --config Release \
        --build_dir build/arm64 \
        --arm64 \
        --ios \
        --ios_toolchain_dir ~/clang-8.0.1_arm64_iOS-12.2/iOS-12.2_arm64 \
        --ios_sysroot ~/clang-8.0.1_arm64_iOS-12.2/iOS-12.2_arm64/SDK/iPhoneOS12.2.sdk \
        --parallel \
        --skip_tests 

**Motivation and Context**
We wanted to deploy libraries with dependencies to onnxruntimefor iOS at our company

### CODE REVIEW

**Comment by tracysh** [comment]
Hi, what was the build error from SgemmKernelNeon.S? I'm guessing you might have seen an undefined symbol because Apple tools prepend an underscore to symbol names. If that's true, then the logic from lib\x86 could be copied over to aarch64: copy asmmacro.h to aarch64, include that and then use C_UNDERSCORE to prepend the needed underscore.

File: onnxruntime/core/providers/cpu/nn/qlinearconv.cc [hunk 1]
```diff
@@ -148,7 +148,7 @@ Status QLinearConv::Compute(OpKernelContext* context) const {
 #endif
 
   const auto* Xdata = X->template Data<uint8_t>();
-  const auto* Wdata = W->template Data<uint8_t>();
+  __attribute__((unused)) const auto* Wdata = W->template Data<uint8_t>();
```
Comment (tracysh) [inline onnxruntime/core/providers/cpu/nn/qlinearconv.cc]: Oops, can you instead replace the expression below with Wdata? The __attribute__ syntax isn't portable to MSVC.

      GemmlowpMultiplyu8u8_u8(**W->template Data<uint8_t>()** + group_id * W_offset,

**Comment by snnn** [comment]
Please report the ONNX issue to https://github.com/onnx/onnx/issues

**Comment by radikalliberal** [comment]
SgemmKernelNeon.S did not compile because ARM ASM and Apple ARM ASM are slightly different.
The ".type" keyword does not exist for apple ARM ASM. I did use the ARM (32bit) fallback with c implementation.

**Comment by tracysh** [comment]
I looked at what OpenBLAS does here, they just do an #ifndef __APPLE__ here, so can you instead try the below. There's still a chance the C_UNDERSCORE change is also needed. Thanks...

        .globl  MlasSgemmKernel\Mode\()
#ifndef __APPLE__
        .type   MlasSgemmKernel\Mode\(),%function
#endif
MlasSgemmKernel\Mode\():

File: cmake/onnxruntime_mlas.cmake [hunk 2]
```diff
@@ -98,6 +98,13 @@ else()
     elseif (CMAKE_ANDROID_ARCH_ABI STREQUAL "x86")
       set(X86 TRUE)
     endif()
+  elseif(CMAKE_SYSTEM_NAME STREQUAL "iOSCross")
+      set(IOS TRUE)
```
Comment (tracysh) [inline cmake/onnxruntime_mlas.cmake]: nit: remove two spaces here

File: tools/ci_build/build.py [hunk 3]
```diff
@@ -373,6 +378,25 @@ def generate_build_tree(cmake_path, source_dir, build_dir, cuda_home, cudnn_home
...
+        arg_names = ["--ios_sysroot <path to sysroot>", "--arm or --arm64", "--ios_toolchain_dir <path to toolchain>"]
+        if any(needed_args):
+            raise BuildError("iOS build canceled due to missing arguments: " + (val + "," for val, cond in zip(arg_names, needed_args) if cond))
+        compilers = sorted(glob.glob(args.ios_toolchain_dir + "/bin/*-clang*"))
+        os.environ["PATH"] = args.ios_toolchain_dir + "/bin:" + os.environ.get("PATH", "")
+        os.environ["LD_LIBRARY_PATH"] = args.ios_toolchain_dir + "/lib:" + os.environ.get("LD_LIBRARY_PATH", "")
+        #print(compilers)
+        #exit(0)
+        if len(compilers) != 2:
+            raise BuildError("error identifiying compilers in toolchaindir")
```
Comment (tracysh) [inline tools/ci_build/build.py]: identifiying->identifying, toolchaindir->toolchain_dir?

**Comment by tracysh** [comment]
Sorry, the comment above was supposed to have `__APPLE__`, but that was converted to bold.

File: tools/ci_build/build.py (hunk 3 above)
Comment (radikalliberal) [reply tools/ci_build/build.py]: corrected typos
I'm not sure if the way I check for the compilers, does work generally for different toolchains.
I assumed that two instances of files matching "/*clang*" do exist in the bin folder of the ios_toochain_dir

**Comment by radikalliberal** [comment]
the change in onnx source files seems only to be necessary because protobuf has been updated for onnxruntime.

File: tools/ci_build/build.py [hunk 4]
```diff
@@ -951,25 +973,21 @@ def main():
...
-                    #Cuda VS extension should be installed to C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\BuildCustomizations
-                    toolset = 'v141,host=x64,version=' + args.msvc_toolset
-                elif args.msvc_toolset:
-                    toolset = 'host=x64,version=' + args.msvc_toolset
-                else:
-                    toolset = 'host=x64'
-                if (args.cuda_version):
-                    toolset += ',cuda=' + args.cuda_version
-                cmake_extra_args = ['-A','x64','-T', toolset, '-G', args.cmake_generator]
+                toolset = 'host=x64'
```
Comment (snnn) [inline tools/ci_build/build.py]: Why do you deleted this section?

Comment (radikalliberal) [reply tools/ci_build/build.py]: bad merge ...

**Comment by tracysh** [comment]
@radikalliberal Did the change to use `__APPLE__` (or more generically `__clang__`) work in SgemmKernelNeon.S?

**Comment by radikalliberal** [comment]
@tracysh no I did not bother using asm for gemm yet. I talked to my supervisor and he suggested i look into it, and if it's done in 2h that's ok but if it costs more effort I won't put more work into it.
I start working on it today and will give you feedback when I'm done.

**Comment by radikalliberal** [comment]
Now SgemmKernelNeon.S is used. It was as simple as you described @tracysh 

The only caveat is that I used protobuf v3.10.1 for this build because of the incompatibility with SetTotalBytesLimit.

File: cmake/onnxruntime_mlas.cmake [hunk 5]
```diff
@@ -123,13 +130,11 @@ else()
     )
   elseif(ARM64)
     enable_language(ASM)
-
```
Comment (tracysh) [inline cmake/onnxruntime_mlas.cmake]: nit: restore missing space?

File: cmake/onnxruntime_mlas.cmake [hunk 6]
```diff
@@ -123,13 +130,11 @@ else()
...
     set(mlas_platform_srcs
       ${ONNXRUNTIME_ROOT}/core/mlas/lib/aarch64/SgemmKernelNeon.S
     )
   elseif(X86)
     enable_language(ASM)
-
```
Comment (tracysh) [inline cmake/onnxruntime_mlas.cmake]: nit: restore missing space?

**Comment by tracysh** [comment]
Thanks for the MLAS changes, that looks good to me. @snnn should review the build script changes.

**Pull request merged**